src/librbd/crypto/luks/EncryptionFormat.cc

```diff
@@ -25,17 +28,38 @@ EncryptionFormat<I>::~EncryptionFormat() {
 
 template <typename I>
 void EncryptionFormat<I>::format(I* image_ctx, Context* on_finish) {
+  std::string passphrase_prefix;
+  if (image_ctx->parent != nullptr) {
+    passphrase_prefix = EncryptionFormat<I>::CLONE_PASSPHRASE_PREFIX;
+  }
   auto req = luks::FormatRequest<I>::create(
-          image_ctx, get_format(), m_alg, std::move(m_passphrase), &m_crypto,
-          on_finish, false);
+          image_ctx, get_format(), m_alg,
+          std::move(passphrase_prefix + m_passphrase), &m_crypto, on_finish,
+          false);
   req->send();
 }
 
 template <typename I>
 void EncryptionFormat<I>::load(I* image_ctx, Context* on_finish) {
+  std::string passphrase_prefix;
```

Comment:
I might be missing something obvious, but why do you have this empty version and why do you re-try using the prefix? It seems like you should know via the image-meta whether or not this image has been formatted using the prefix. 

Also, while I am thinking about it, what happens when you flatten a previously formatted image? We should probably remove the parent cryptor but due to the prefix, it's not going to be able to be opened by dm-crypt (unless you know the magic prefix value).

Reply:
1. Problem is image-meta is cloned. So when I open up an image I do not know if it was formatted (at least not until I check if the parent has the same metadata, which I currently don't). Changing this will require changing the API of `EncryptionFormat.load(...)` to include an extra parameter like `is_formatted`.

2. Since it's only the wrapping passphrase changed, we can change the passphrase upon flattening, removing the prefix.

Reply:
We can tweak the clone operation to not copy certain image-meta keys which I think would be a clean solution. We could do it generically via a unique prefix to indicate that it shouldn't be copied (i.e. `.rbd_bound_XYZ`) or we can just create a list of known internal keys that shouldn't be copied.

As for the flatten, does that mean that you would need to keep the passphrase in-memory just in case a flatten operation might need to change it in the LUKS key slot?

Reply:
So from what I understand today:
`.rbd` prefix is NOT copied on clone/copy.
We want our key to be copied always, except for clones.
I think that adding a another prefix `.rbd_bound` with this semantics is confusing, assuming the copy semantics of `.rbd`.
@trociny What do you think?

Reply:
May be skip all `.rbd` prefixed keys in copy case too (as it is now), but after creating the target image explicitly add the keys you want here?

Reply:
i.e. to explicitly add the keys on both `DeepCopyRequest.cc` and `api\Migration.cc`?
These are the current places I see that use `MetadataCopyRequest` (except for `CloneRequest`)

Reply:
And I believe in `internal.cc:copy`?

The idea is when copying/migrating an image the first step we do is to create the destination image with the property we want (objects size, feature, etc...) -- they may be different from the source image properties. So we are doing this explicitly. I don't see reasons why we shouldn't do the same with the encryption format. I can imagine a situation when in future (if not from now) we will want to be able to use a different encryption format for the destination image?

Reply:
looks like `internal.cc:copy` is doing a shallow copy, and that the destination image is not created with the same parent as the source. Am I right?
Is there a sense to make that kind of copy for a cloned image?

Reply:
Ah, indeed. I think I misunderstood was going here. I should admit I have not had a chance to read the code yet, only glanced though it.

Yes, you don't need to do this for "flat" copy. But this also confirms that is a not a very good idea to modify `MetadataCopyRequest` to copy these encryption keys, because we don't need to copy them for "flat" copy, right?

So I think I still like the approach to skip all internal metadata keys when copying, and then explicitly add/copy those that we need. But right now I don't have a strong opinion about this as I am still not very familiar with this code.

Reply:
I'm thinking adding a new 'bool include_encryption_keys` to `GetMetadataRequest` and `MetadataCopyRequest`.
This will require all use-cases to decide for themselves, and also save me of duplicating code to create something like `EncryptionMetadataCopyRequest`.

Reply:
LGTM. My suggestion is just to do this explicitly (at least for now), i.e. skip(don't skip) the keys you want exactly instead of using generic `.rbd_bound` prefix.